### PR TITLE
[vLLM] batch paged_fill_cache across users in prefill (compile/perf improvement)

### PR DESCRIPTION
**Pull request by kmabeeTT**
### Ticket

- Closes/relates to [#4880](https://github.com/tenstorrent/tt-xla/issues/4880)
- Depends on [tt-mlir #8613](https://github.com/tenstorrent/tt-mlir/pull/8613) (verifier relaxation) and [tt-metal #45117](https://github.com/tenstorrent/tt-metal/pull/45117) (kernel relaxation); both arrive via regular tt-metal -> tt-mlir -> tt-xla uplift.

### Problem

- Motivated by **compile time and prefill performance**. `_handle_paged_attention`'s prefill branch ran a Python `for batch_idx in range(inputs.users):` loop, calling `paged_fill_cache` once per user per K/V per layer. At trace time the loop unrolls into the graph, so b32 with 32 layers produced 2048 `paged_fill_cache` ops plus ~7000 supporting `slice` / `to_device` / `get_device` ops in the compiled prefill graph (vs. 64 at b1), driving both engine init and per-prefill latency.

### What's changed

- Single hunk in `integrations/vllm_plugin/vllm_tt/attention.py` (+18 / -18). The per-user loop is replaced with one batched call per K/V: `batch_idxs = torch.arange(key_for_update.shape[0], dtype=int32, ...)` produces an N-element `batch_idx_tensor` and the op is invoked once with the full batched K/V instead of N times with per-row slices.

- b1 behavior is unchanged (`arange(1)` collapses to the same single-row call the old loop produced). Multi-user prefill is what changes.

### Testing

End-to-end coverage is unchanged:
- `tests/integrations/vllm_plugin/generative/test_kv_cache_bleed.py` exercises 4-way concurrent prefill (`max_num_seqs=4`) on TinyLlama and asserts no cross-user contamination.
- `tests/benchmark/test_vllm_benchmarks.py` Llama-3.2-3B b32 sweep is the perf source of truth.

Headlines from the b32 sweep on p150 (full numbers and raw logs in the follow-up comment):

- **Wall-clock 2–3x faster**: seq128 153 s -> 70 s, seq1024 476 s -> 147 s, seq2048 610 s -> 196 s.
- **Samples/sec is now flat across seq_len**: was 16.3 / 17.1 / 18.5 at seq 2048 / 1024 / 128 (dropping with longer seq because the compiled graph grew with `seq_len`-dependent fan-out); now ~19.0–19.3 at all three.
- Most of the wall-clock win is engine init (compile + warmup); residual scaling with `max_model_len` comes from the unrelated exponential bucket compile (`[1, 32, 64, ..., max]`).

### Checklist

- [x] Existing tests provide coverage for changes
- [x] [tt-mlir #8613](https://github.com/tenstorrent/tt-mlir/pull/8613) + [tt-metal #45117](https://github.com/tenstorrent/tt-metal/pull/45117) merged and uplifted to tt-xla before this lands. Will hold off from merging this until changes arrive here.
- [ ] vLLM Nightly: https://github.com/tenstorrent/tt-xla/actions/runs/27310498979

### CODE REVIEW

**Review by AleksKnezevic — APPROVED** [review]

**Comment by kmabeeTT** [comment]
### Extra Info For Reviewers

Full perf numbers + raw log excerpts. 

## Numbers, original metric — Llama-3.2-3B b32, single chip (p150), `test_vllm_benchmarks.py`

`Sample per second` here is the pre-#4953 metric (includes engine init / compile / prefill in the denominator):

| seq_len | wall-clock (before) | wall-clock (after) | speedup | samples/sec (before) | samples/sec (after) |
|---:|---:|---:|---:|---:|---:|
| 128  | 153.37 s | **70.16 s**  | **2.19x** | 18.54 | 19.03 |
| 1024 | 475.66 s | **146.84 s** | **3.24x** | 17.10 | 19.03 |
| 2048 | 610.16 s | **196.01 s** | **3.11x** | 16.32 | 19.31 |

Observations:

- **Throughput flattens.** On `main`, samples/sec dropped with longer `seq_len` (16.32 / 17.10 / 18.54 at 2048 / 1024 / 128) because the compiled graph grew with `seq_len`-dependent op fan-out (2048 `paged_fill_cache` ops at b32). After the fix it sits at ~19.0–19.3 across all three.
- **Wall-clock dominated by engine init.** The samples/sec deltas are modest (~3–18%); the headline 2–3x wall-clock speedup is overwhelmingly compile + warmup.
- **Residual wall-clock scaling with `max_model_len` is unrelated** to this change. It comes from the exponential prefill bucket list in `model_runner.py` (`[1, 32, 64, 128, ..., max]`, each bucket a separate graph compile, ~25–30 s each). seq=128 compiles 4 buckets, seq=2048 compiles 8.

Raw logs:

Before (`main` @ 44a614192):

```
llama-3.2-3b-batch32_seq128.log:| Sample per second: 18.538099791161347
llama-3.2-3b-batch32_seq128.log:================== 1 passed, 18 warnings in 153.37s (0:02:33) ==================
llama-3.2-3b-batch32_seq1024.log:| Sample per second: 17.103699472879022
llama-3.2-3b-batch32_seq1024.log:================== 1 passed, 18 warnings in 475.66s (0:07:55) ==================
llama-3.2-3b-batch32_seq2048.log:| Sample per second: 16.315267367469527
llama-3.2-3b-batch32_seq2048.log:================== 1 passed, 18 warnings in 610.16s (0:10:10) ==================
```

After (this PR):

```
llama-3.2-3b-batch32_seq128.log:| Sample per second: 19.027504852132070
llama-3.2-3b-batch32_seq128.log:================== 1 passed, 18 warnings in  70.16s (0:01:10) ===================
llama-3.2-3b-batch32_seq1024.log:| Sample per second: 19.027511044887994
llama-3.2-3b-batch32_seq1024.log:================== 1 passed, 18 warnings in 146.84s (0:02:26) ==================
llama-3.2-3b-batch32_seq2048.log:| Sample per second: 19.313553916710706
llama-3.2-3b-batch32_seq2048.log:================== 1 passed, 18 warnings in 196.01s (0:03:16) ==================
```

## Numbers, decode-only metric — after [#4953](https://github.com/tenstorrent/tt-xla/pull/4953)

[#4953](https://github.com/tenstorrent/tt-xla/pull/4953) fixed `Sample per second` to exclude prefill / engine-init time from the denominator, so this number now reflects steady-state decode throughput rather than end-to-end wall-clock. Re-running the same b32 sweep on `main` and on this PR with the fixed metric:

| seq_len | decode samples/sec (before) | decode samples/sec (after) | delta |
|---:|---:|---:|---:|
| 128  | 21.76 | **22.41** | **+3.0%** |
| 2048 | 18.58 | **22.36** | **+20.3%** |

Why decode throughput moves at all (when the op being changed is in *prefill*): the per-user `paged_fill_cache` loop wasn't only inflating the compiled prefill graph — at long `max_model_len` it was also dragging steady-state decode throughput down (cache locality, program-cache pressure, or both — not chased to a precise root cause). With a single batched call, decode at seq=2048 lifts from 18.58 to 22.36 (+20%), matching the seq=128 case and confirming the residual was attributable to this codepath.

Raw logs:

```
// without fix (main @ 44a614192, post-#4953 metric)
perf_bench_may26/44a614192_jona_change/llama-3.2-3b-batch32_seq128.log:|  Sample per second: 21.76450836588681
perf_bench_may26/44a614192_jona_change/llama-3.2-3b-batch32_seq2048.log:| Sample per second: 18.576465055477083

// with fix (this PR, post-#4953 metric)
perf_bench_may26/batch_paged_fill_cache_issue_4880_poc_jona_change/llama-3.2-3b-batch32_seq128.log:|  Sample per second: 22.409910275390313
perf_bench_may26/batch_paged_fill_cache_issue_4880_poc_jona_change/llama-3.2-3b-batch32_seq2048.log:| Sample per second: 22.357091173812766
```

File: integrations/vllm_plugin/vllm_tt/attention.py [hunk 1]
```diff
@@ -475,27 +475,27 @@ def _handle_paged_attention(
...
-                )
-                v_cache = torch.ops.tt.paged_fill_cache(
-                    v_cache,
-                    value_for_update[batch_idx : batch_idx + 1],
-                    attn_metadata.fill_page_table,
-                    batch_idx=torch.tensor(
-                        [batch_idx], dtype=torch.int32, device=v_cache.device
-                    ),
-                )
+            batch_idxs = torch.arange(
```
Comment (ssaliceTT) [inline integrations/vllm_plugin/vllm_tt/attention.py:492]: @AleksKnezevic would [this](https://github.com/tenstorrent/tt-xla/pull/5108#discussion_r3377077308) comment you'd mentioned apply here?

Comment (AleksKnezevic) [reply integrations/vllm_plugin/vllm_tt/attention.py:492]: yes it would, good catch.

Comment (kmabeeTT) [reply integrations/vllm_plugin/vllm_tt/attention.py:492]: thx for flagging, discussed w/ Asif offline, he has a change in mind but not quite ready, prefer to land this as-is.  I left todo comment here and opened issue for follow up: https://github.com/tenstorrent/tt-xla/issues/5154

**Comment by kmabeeTT** [comment]
I thought I could be smart and temporarily include a fix for broken build on main just to get onPR to run (it would drop naturally once it got merged through it's own PR), but forgot it expanded reviewer list (whoops), so I dropped it from here now (apologies for reviewer spam, it's just limited to VLLM group now).

**Review by mmanzoorTT — APPROVED** [review]

**Comment by kmabeeTT** [comment]
tt-mlir uplift w/ support added, re-tested locally with and without fix, we continue to see expected improvements.

Edit: Lower numbers here since I had trace disabled.

 **Perf qualification** (single P150, b32, opt=1, trace off, bf8 KV; before = per-user loop, after = batched):
 
 Batching the per-user `paged_fill_cache` into one call cuts prefill **compile time by 47–77%** (Llama-3.2-3B @2k: 698s→158s; Llama-3.1-8B @2k: 873s→230s — the larger the seq, the bigger the win), with **+5–20% samples/sec** and **−4–8% TTFT** across {3B, 8B} × {128, 1k, 2k}. No correctness change (identical binary; only the prefill-fill path differs).


  | model | seq | compile s (before→after) | samples/sec (before→after) | TTFT ms (before→after) |
  |---|---|---|---|---|
  | Llama-3.2-3B | 128 | 169.7 → 89.7 (−47%) | 13.77 → 14.70 (+7%) | 624.2 → 586.0 (−6%) |
  | Llama-3.2-3B | 1024 | 551.0 → 135.4 (−75%) | 12.11 → 14.29 (+18%) | 639.1 → 585.1 (−8%) |
  | Llama-3.2-3B | 2048 | 698.0 → 158.1 (−77%) | 12.13 → 14.31 (+18%) | 634.3 → 585.7 (−8%) |
  | Llama-3.1-8B | 128 | 217.2 → 106.5 (−51%) | 11.95 → 12.57 (+5%) | 1091.1 → 1046.7 (−4%) |
  | Llama-3.1-8B | 1024 | 684.0 → 199.4 (−71%) | 10.72 → 12.32 (+15%) | 1099.6 → 1048.1 (−5%) |
  | Llama-3.1-8B | 2048 | 872.9 → 230.3 (−74%) | 10.41 → 12.48 (+20%) | 1103.7 → 1045.2 (−5%) |
  
  Unrelated comparison of trace off vs on w/ fix just since I had it:
  
  | model | seq | samples/sec (trace off → on) |
  |---|---|---|
  | Llama-3.2-3B | 128 | 14.70 → 25.84 (+76%) |
  | Llama-3.2-3B | 1024 | 14.29 → 25.59 (+79%) |
  | Llama-3.2-3B | 2048 | 14.31 → 25.42 (+78%) |
  | Llama-3.1-8B | 128 | 12.57 → 16.71 (+33%) |
  | Llama-3.1-8B | 1024 | 12.32 → 16.65 (+35%) |
  | Llama-3.1-8B | 2048 | 12.48 → 16.61 (+33%) |
